File: codility/MaxCounters/MaxCounters.java
/**
 * Author: Liang Lin
 * Email: lianglin@outlook.com
 * Date: 2016-03-27
 */
 
Class Solution {
    public int solution(int N, int[] A) {
        int mc = 0;
        int lmc = 0;
        int n = A.length;
        int[] B = new int[N];
        for(int i=0; i<n; i++) {
            if(N + 1 == A[i]) {
                lmc = mc;
            } else {
                B[A[i]-1] = Math.max(lmc+1, B[A[i]-1]+1);
                mc = Math.max(mc, B[A[i]-1]);
            }
        }
        int m = B.length;
        for(int j=0; j<m; j++) {
            B[j] = Math.max(lmc, B[j]);
        }
        return B;
    }
}

File: codility/4.CountingElements/MaxCounters/MaxCounters.py
# Author: Liang Lin
# Email: lianglin@outlook.com
# Date: 2016-03-27
class Solution:
  def solution(self, N, A):
    #max counter
    mc = 0
    lmc = 0
    '''
    :type B: List[int]
    :rtype: int
    '''
    B = [0] * N
    for i, val in enumerate(A):
      if val == N + 1:
        lmc = mc
      else:
        B[val-1] = max(lmc+1, B[val-1]+1)
        mc = max(mc, B[val-1])

    for i, val in enumerate(B):
      B[i] = max(lmc, val)
    return B
sol = Solution()
print(sol.solution(3,[2,1,1,2,3,4,1,2,3,2,2]))
      